File: examples/pdu_communication/src/asset_plant.c
#include "hako_asset.h"
#include "hako_conductor.h"
#include "pdu_info.h"
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

static int my_on_initialize(hako_asset_context_t* context)
{
    return 0;
}
static int my_on_reset(hako_asset_context_t* context)
{
    return 0;
}

static int my_on_manual_timing_control(hako_asset_context_t* context)
{
    Hako_Twist motor;
    Hako_Twist pos;
    printf("INFO: on_manual_timing_control enter\n");
    int result = 0;
    double count = 0;
    while (result == 0) {
        pos.linear.x = count + 1;
        pos.linear.y = count + 2;
        pos.linear.z = count + 3;
        int ret = hako_asset_pdu_write("Robot", PDU_POS_CHANNEL_ID, (const char*)(&pos), sizeof(pos));
        if (ret != 0) {
            printf("ERROR: hako_asset_pdu_read erro: %d\n", ret);
        }
        result = hako_asset_usleep(1000);
        if (result != 0) {
            break;
        }

        ret = hako_asset_pdu_read("Robot", PDU_MOTOR_CHANNEL_ID, (char*)(&motor), sizeof(motor));
        if (ret != 0) {
            printf("ERROR: hako_asset_pdu_read erro: %d\n", ret);
        }
        printf("%llu: motor data(%f, %f, %f)\n", hako_asset_simulation_time(), motor.linear.x, motor.linear.y, motor.linear.z);


        result = hako_asset_usleep(1000);
        usleep(1000*1000);
        count++;
    }
    printf("INFO: on_manual_timing_control exit\n");
    return 0;
}

static hako_asset_callbacks_t my_callback = {
    .on_initialize = my_on_initialize,
    .on_manual_timing_control = my_on_manual_timing_control,
    .on_simulation_step = NULL,
    .on_reset = my_on_reset
};
int main(int argc, const char* argv[])
{
    if (argc != 2) {
        printf("Usage: %s <config_path>\n", argv[0]);
        return 1;
    }
    const char* asset_name = "Writer";
    const char* config_path = argv[1];
    hako_time_t delta_time_usec = 1000000;

    hako_conductor_start(delta_time_usec, delta_time_usec);
    int ret = hako_asset_register(asset_name, config_path, &my_callback, delta_time_usec, HAKO_ASSET_MODEL_PLANT);
    if (ret != 0) {
        printf("ERORR: hako_asset_register() returns %d.", ret);
        return 1;
    }
    ret = hako_asset_start();
    printf("INFO: hako_asset_start() returns %d\n", ret);

    hako_conductor_stop();
    return 0;
}

File: examples/pdu_communication/asset_plant.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import hakopy
import hako_pdu
import pdu_info
import sys
import time

def my_on_initialize(context):
    return 0

def my_on_reset(context):
    return 0

pdu_manager = None
def my_on_manual_timing_control(context):
    global pdu_manager
    print("INFO: on_manual_timing_control enter")
    pdu_pos = pdu_manager.get_pdu('Robot', pdu_info.PDU_POS_CHANNEL_ID)
    pdu_motor = pdu_manager.get_pdu('Robot', pdu_info.PDU_MOTOR_CHANNEL_ID)
    pos = pdu_pos.get()
    result = True
    count = 0
    while result:
        pos['linear']['x'] = count + 1
        pos['linear']['y'] = count + 2
        pos['linear']['z'] = count + 3
        
        ret = pdu_pos.write()
        if ret == False:
            print('"ERROR: hako_asset_pdu_write')
            break
        result = hakopy.usleep(1000)
        if result == False:
            break

        motor = pdu_motor.read()
        if motor == None:
            print('ERROR: hako_asset_pdu_read')
        print(f'{hakopy.simulation_time()}: motor data({motor['linear']['x']}, {motor['linear']['y']}, {motor['linear']['z']})')

        result = hakopy.usleep(1000)
        if result == False:
            break
        count = count + 1
        time.sleep(1)  # 1秒スリープ
    print("INFO: on_manual_timing_control exit")
    return 0

my_callback = {
    'on_initialize': my_on_initialize,
    'on_simulation_step': None,
    'on_manual_timing_control': my_on_manual_timing_control,
    'on_reset': my_on_reset
}
def main():
    global pdu_manager
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <config_path>")
        return 1

    asset_name = 'Writer'
    config_path = sys.argv[1]
    delta_time_usec = 1000000

    pdu_manager = hako_pdu.HakoPduManager('/usr/local/lib/hakoniwa/hako_binary/offset', config_path)

    hakopy.conductor_start(delta_time_usec, delta_time_usec)
    ret = hakopy.asset_register(asset_name, config_path, my_callback, delta_time_usec, hakopy.HAKO_ASSET_MODEL_PLANT)
    if ret == False:
        print(f"ERROR: hako_asset_register() returns {ret}.")
        return 1

    ret = hakopy.start()
    print(f"INFO: hako_asset_start() returns {ret}")

    hakopy.conductor_stop()
    return 0

if __name__ == "__main__":
    sys.exit(main())
